### PR TITLE
interactive-canvas v1

### LINKED ISSUE #9: Implement initial render of statement network
For now, we just want to implement a quick initial render so we can start gathering feedback.

My suggestion for rendering them is as follows:
- Render each statement in the same (horizontal) format, one above the other
- For each statement, 'reserve' a rectangle of space.
- For each component of the statement, plot it using a fixed relative offset from the top-left corner of the reserved rectangle.

For now, we can simply draw each component to be 'oversized' to make sure it fits all text

### PR DESCRIPTION
**Pull request by sjvrijn**
closes #9 

Adds an interactive canvas to display (a first draft of) the statements

### CODE REVIEW

File: includes/chart.js [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+
+            // Add text inside the shape
+            var text = document.createElementNS("http://www.w3.org/2000/svg", "text");
+            text.textContent = row[template_node.type];
+            text.setAttribute("fill", "white");
+            text.setAttribute("font-size", "12px");
+            text.setAttribute("font-family", "calibri");
+            text.setAttribute("text-anchor", "middle"); // Center align the text
+
+            var textX = x, textY = y;
```
Comment (OleMussmann) [inline includes/chart.js:146]: `textX` and `textY` could be more descriptive, I don't know what they mean.

File: includes/chart.js [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+            text.setAttribute("x", textX);
+            text.setAttribute("y", textY);
+
+            group.appendChild(newNode);
+            group.appendChild(text);
+            svg.appendChild(group);
+
+            // Add click event to display the clicked node
+            newNode.addEventListener('click', function () {
+                document.getElementById("clicked-node").textContent = "Clicked Node: " + `${row.id}_${template_node.id}`;
```
Comment (OleMussmann) [inline includes/chart.js:167]: The ID "clicked-node" could maybe named after it's function/role, not after what happens if you do something outside of this `div`. Maybe "node-logging"? I don't know enough about its function to come up with a good one. :-)

File: includes/chart.js [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+    "indirectObject",
+    "typeOfIndirectObject",
+    "activationCondition",
+    "executionConstraint",
+    "orElse"
+];
+
+function addNodesAndLinks(rowValues) {
+    var svgContainer = document.getElementById("svgContainer");
+    var svg = svgContainer.querySelector("svg");
```
Comment (OleMussmann) [inline includes/chart.js:80]: Most of the `var` declarations in this PR could probably be `let`'s. I don't know enough about web dev, but I would imagine it to be safer to limit the scope to where it's needed. I'll skip the following ones, so please check where it's applicable.

https://stackoverflow.com/questions/762011/what-is-the-difference-between-let-and-var

File: includes/chart.js [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+            if (!row[template_node.type]) {
+                return;
+            }
+
+            var group = document.createElementNS("http://www.w3.org/2000/svg", "g");
+            var newNode;
+            var x = template_node.x + rowX;
+            var y = template_node.y + rowY + (i * rowHeight);
+            if (template_node.shape === 'polygon') {
```
Comment (OleMussmann) [inline includes/chart.js:101]: There are quite some if/else toggles for the different node shapes. I could imagine it to be more extensible and performant if these happen once (or at least less often). I would need to think about it (or maybe co-work for a bit with you) to see how that can be simplified.

The sniff check says it's quite some code repetition that could probably be avoided.

File: includes/chart.js [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+        if (sourceNode.tagName === 'ellipse') {
+            sourceX = parseFloat(sourceNode.getAttribute("cx"));
+            sourceY = parseFloat(sourceNode.getAttribute("cy"));
+        } else if (sourceNode.tagName === 'polygon') {
+            // Get the bounding box of the polygon
+            var sourceBBox = sourceNode.getBoundingClientRect();
+            // Define an offset to adjust the initial edge position (empirical adjustment)
+            var offsetX = 275; // Adjust this value as needed (!!! This is a little temporary hack !!!)
+            var offsetY = 225; // Adjust this value as needed (!!! This is a little temporary hack !!!)
```
Comment (OleMussmann) [inline includes/chart.js:255]: You won't find these lines anymore after a few weeks. ;-) Please use variables for this instead of hard-coding.

File: includes/chart.js [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+    var numPoints = points.numberOfItems;
+    var x = 0, y = 0;
+
+    for (var i = 0; i < numPoints; i++) {
+        x += points.getItem(i).x;
+        y += points.getItem(i).y;
+    }
+
+    x /= numPoints;
+    y /= numPoints;
```
Comment (OleMussmann) [inline includes/chart.js:316]: Why is this happening? Descriptive var names and occasionally more verbosity could help readability.

File: includes/chart.js [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+}
+
+
+// Function to highlight node1
+function highlightNode(nodeID) {
+    var node1 = document.getElementById(nodeID);
+    if (node1.getAttribute("stroke") === "green") {
+        node1.removeAttribute("stroke");
+    } else {
+        node1.setAttribute("stroke", "green");
```
Comment (OleMussmann) [inline includes/chart.js:328]: styles, css vars, see above

File: includes/header.php [hunk 8]
```diff
@@ -198,6 +140,18 @@
                                 </div>
                             </a>
 
+                            <a class="nav-link">
+                                <div class="sb-nav-link-icon">
+                                    <button type="button" class="btn btn-outline-primary" onclick="loadDemo()">
+                                        <!-- Add upload svg icon -->
+                                        <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-file-earmark-arrow-up-fill" viewBox="0 0 16 16">
+                                        <path d="M9.293 0H4a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h8a2 2 0 0 0 2-2V4.707A1 1 0 0 0 13.707 4L10 .293A1 1 0 0 0 9.293 0M9.5 3.5v-2l3 3h-2a1 1 0 0 1-1-1M6.354 9.854a.5.5 0 0 1-.708-.708l2-2a.5.5 0 0 1 .708 0l2 2a.5.5 0 0 1-.708.708L8.5 8.707V12.5a.5.5 0 0 1-1 0V8.707z"/>
+                                        </svg>
```
Comment (OleMussmann) [inline includes/header.php:149]: I would probably use pre-made icons instead of SVGs for this. What if you want to change the style later? The `path` data is not really tweakable by humans.

You can use those, for example: https://fontawesome.com/

File: includes/chart.js [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+                        // Update text position for polygon
+                        text.setAttribute("x", parseFloat(text.getAttribute("x")) + deltaX);
+                        text.setAttribute("y", parseFloat(text.getAttribute("y")) + deltaY);
+
+                    } else if (newNode.tagName === 'rect') {
+                        newNode.setAttribute("x", event.clientX - offsetX);
+                        newNode.setAttribute("y", event.clientY - offsetY);
+                        // Update text position for rectangle
+                        text.setAttribute("x", parseFloat(newNode.getAttribute("x")) + parseFloat(newNode.getAttribute("width")) / 2);
+                        text.setAttribute("y", parseFloat(newNode.getAttribute("y")) + parseFloat(newNode.getAttribute("height")) / 2);
```
Comment (OleMussmann) [inline includes/chart.js:213]: Do you have a linter for line length? More than 100something chars is quite long.

File: includes/header.php [hunk 10]
```diff
@@ -220,22 +174,15 @@
...
-                    <div id='column_names' hidden></div>
-                    <hr>
-                    <div id='rows_values' hidden></div>
-
-                    
                 </div>
                 <div class="modal-footer">
                     <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
-                    <button type="button" class="btn btn-primary" data-bs-dismiss="modal" onclick = 'confirmUpload()'>Confirm upload</button>
+                    <button type="button" class="btn btn-primary" data-bs-dismiss="modal" onclick = 'confirmUpload("", "")'>Confirm upload</button>
```
Comment (OleMussmann) [inline includes/header.php:180]: `uploader.js`, where `confirmUpload` is defined (and apparently has changed) is not part of this PR. Is this a fix for something that broke during refactoring? Or should more files be included in this PR?

File: index.php [hunk 11]
```diff
@@ -25,8 +29,27 @@
...
+                                        max-height: 60vh; /* Adjust as needed */
+                                    }
 
-                                
+                                    /* Added styles for the inner svg */
+                                    #svgContainer svg {
+                                        min-width: 1200px; /* Set a minimum width to enable horizontal scrolling */
+                                        min-height: 1600px; /* Set a minimum height to enable vertical scrolling */
+                                    }
+                                </style>
```
Comment (OleMussmann) [inline index.php:48]: This should probably be in an external CSS file

File: index.php [hunk 12]
```diff
@@ -25,8 +29,27 @@
...
+
+                                <hr>
+                                <div id="clicked-node"></div>
```
Comment (OleMussmann) [inline index.php:51]: ID could have a different name, see above

**Review by OleMussmann — COMMENTED** [review]
Nice functionality.

My main hints for improvement would be about hard-coding styles/colors that could probably be set as variables, and repeating code sections due to the if/else blocks for the different shapes. The latter would need need some re-thinking about the architecture.

Since it works like it is (and this would be a larger refactoring) I would suggest to put that as a separate issue and fix it in a later PR.

File: includes/chart.js (hunk 2 above)
Comment (sjvrijn) [reply includes/chart.js:171]: This specific `clicked-node` div is a development leftover and currently serves as a placeholder for eventually highlighting the relevant row in the table when a node/statement gets selected.

File: index.php (hunk 11 above)
Comment (sjvrijn) [reply index.php:48]: Deferred to #23

File: includes/chart.js [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+    links: [
+        { source: 1, target: 2 },
+        { source: 2, target: 3 },  // include deontic info here
+        { source: 3, target: 4 },
+        { source: 5, target: 3 },
+        { source: 4, target: 6 },
+    ],
+};
+
+const colors = { formal: '#99cc00', informal: '#f4b183' }
```
[the remark this answers is not in the document]
Comment (sjvrijn) [reply includes/chart.js:61]: We'll defer this to #23

File: includes/chart.js [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+                newNode.setAttribute("height", template_node.height); // Use the height attribute
+                if (template_node.rx && template_node.ry) {
+                    newNode.setAttribute("rx", template_node.rx); // Set horizontal radius for rounded corners
+                    newNode.setAttribute("ry", template_node.ry); // Set vertical radius for rounded corners
+                }
+            }
+
+            newNode.setAttribute("id", `${row.id}_${template_node.id}`);
+            newNode.setAttribute("fill", colors[row.statementType]);
+            newNode.setAttribute("stroke", "black")
```
[the remark this answers is not in the document]
Comment (sjvrijn) [reply includes/chart.js:136]: Deferred to #23

File: includes/chart.js (hunk 7 above)
Comment (sjvrijn) [reply includes/chart.js:332]: Deferred to #23

File: includes/chart.js (hunk 9 above)
Comment (sjvrijn) [reply includes/chart.js:212]: Not yet, this is an open issue #18

File: includes/header.php (hunk 8 above)
Comment (sjvrijn) [reply includes/header.php]: Done

File: includes/chart.js (hunk 4 above)
Comment (sjvrijn) [reply includes/chart.js:101]: All common attributes are assigned outside of the if-else ifs, the code blocks are the slightly different actions associated with each shape. So as far as I can see, there is nothing that can be explicitly de-duplicated...

I've made a separate issue to revisit this (#24) to see if there's any viable refactoring to be done, that way we can at least close this PR and move on with smaller changes from now on.

File: includes/chart.js [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+        row_template.nodes.forEach(function (template_node) {
+
+            if (!row[template_node.type]) {
+                return;
+            }
+
+            var group = document.createElementNS("http://www.w3.org/2000/svg", "g");
+            var newNode;
+            var x = template_node.x + rowX;
+            var y = template_node.y + rowY + (i * rowHeight);
```
[the remark this answers is not in the document]
Comment (sjvrijn) [reply includes/chart.js]: Yes, have renamed to nodeX/Y to make it clearer

File: includes/chart.js (hunk 1 above)
Comment (sjvrijn) [reply includes/chart.js]: They are the x and y coordinate of where to render the text in a node. Do you have any suggestions for naming it to make it clearer? I can't come up with anything

File: includes/chart.js (hunk 6 above)
Comment (sjvrijn) [reply includes/chart.js]: x/y names made more explicit

File: includes/chart.js [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+    "executionConstraint",
+    "orElse"
+];
+
+function addNodesAndLinks(rowValues) {
+    var svgContainer = document.getElementById("svgContainer");
+    var svg = svgContainer.querySelector("svg");
+
+    const rowHeight = 180;
+    const rowX = 75, rowY = 30;
```
[the remark this answers is not in the document]
Comment (sjvrijn) [reply includes/chart.js]: renamed to `startDrawingAt.x/y`

File: includes/chart.js (hunk 5 above)
Comment (sjvrijn) [reply includes/chart.js]: Deferred to #25

File: includes/chart.js [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+            newNode.setAttribute("id", `${row.id}_${template_node.id}`);
+            newNode.setAttribute("fill", colors[row.statementType]);
+            newNode.setAttribute("stroke", "black")
+
+            // Add text inside the shape
+            var text = document.createElementNS("http://www.w3.org/2000/svg", "text");
+            text.textContent = row[template_node.type];
+            text.setAttribute("fill", "white");
+            text.setAttribute("font-size", "12px");
+            text.setAttribute("font-family", "calibri");
```
[the remark this answers is not in the document]
Comment (sjvrijn) [reply includes/chart.js]: Deferred to #23

**Review by sjvrijn — COMMENTED** [review]
@OleMussmann Thanks for your review! I believe I've addressed most comments, either directly or by deferring them to a specific issue.
I'll be merging this PR now given the time constraints of those involved and so we can move forward by working in smaller increments
